kernel/src/fs/sysfs/inode.rs

```diff
@@ -0,0 +1,827 @@
+        // This freedom is guaranteed by Linux as documented in
+        // [the man page of getdents](https://man7.org/linux/man-pages/man2/getdents.2.html).
+        //
+        // Our implementation of sysfs interprets the `offset`
+        // as an _inode number_.
+        // By inode numbers, directory entries will have a _stable_ order
+        // across different calls to `readdir_at`.
+        // The `new_dentry_iter` is responsible for filtering out entries
+        // with inode numbers less than `start_ino`.
+        let start_ino = offset as Ino;
+        let mut count = 0;
+        let mut last_ino = start_ino;
+
+        let mut iter = self.new_dentry_iter(start_ino + 1);
+
+        while let Some(dentry) = iter.next() {
+            // The offset reported back to the caller should be the absolute position
+            let next_offset = (dentry.ino + 1) as usize;
+            let res = visitor.visit(&dentry.name, dentry.ino, dentry.type_, next_offset);
+
+            if res.is_err() {
+                if count == 0 {
+                    return Err(Error::new(Errno::EINVAL));
+                } else {
+                    break;
+                }
+            }
+            count += 1;
+            last_ino = dentry.ino;
+        }
+
+        if count == 0 {
+            Ok(0)
+        } else {
+            // Return absolute offset instead of an increment
+            Ok((last_ino + 1) as usize)
+        }
+    }
+
+    fn read_link(&self) -> Result<String> {
+        match &self.inner_node {
+            InnerNode::Symlink(s) => Ok(s.target_path().to_string()),
+            _ => Err(Error::new(Errno::EINVAL)),
+        }
+    }
+
+    fn write_link(&self, _target: &str) -> Result<()> {
+        Err(Error::new(Errno::EPERM))
+    }
+
+    fn as_device(&self) -> Option<Arc<dyn Device>> {
+        None
+    }
+
+    fn ioctl(&self, _cmd: IoctlCmd, _arg: usize) -> Result<i32> {
+        Err(Error::new(Errno::ENOTTY))
+    }
+
+    fn sync_all(&self) -> Result<()> {
+        Ok(())
+    }
+
+    fn sync_data(&self) -> Result<()> {
+        Ok(())
+    }
+
+    fn fallocate(&self, _mode: FallocMode, _offset: usize, _len: usize) -> Result<()> {
+        Err(Error::new(Errno::EOPNOTSUPP))
+    }
+
+    fn poll(&self, mask: IoEvents, _poller: Option<&mut PollHandle>) -> IoEvents {
+        let mut events = IoEvents::empty();
+        if let InnerNode::Attr(attr, _) = &self.inner_node {
+            if attr.flags().contains(SysAttrFlags::CAN_READ) {
+                events |= IoEvents::IN;
+            }
+            if attr.flags().contains(SysAttrFlags::CAN_WRITE) {
+                events |= IoEvents::OUT;
+            }
+        }
+        events & mask
+    }
+
+    fn is_dentry_cacheable(&self) -> bool {
+        true
+    }
+}
+
+/// A unified iterator for sysfs directory entries.
+/// Iterates through attributes, then child nodes (if any), then "." and "..".
+struct SysFsDentryIter<'a> {
```

Comment:
This type is needless. You can do `iter1.chain(iter2).chain(iter3)`, which basically does the same job as `SysFsDentryIter`. The return type of `new_dentry_iter` can be changed to `impl Iterator<Item = Dentry>`.